### ISSUE TITLE
Default CFLAGS make flint incompatible with new ntl versions

### ISSUE THREAD

**Issue by timokau** [wbhart/flint2#487]
Since ntl 11.0.0, ntl compiles with `-std=c++11` by default. If `CFLAGS` is unset, flint defaults to `-ansi`:
https://github.com/wbhart/flint2/blob/trunk/configure#L615
That makes flint and ntl incompatible (assuming default configuration for both). I think flint shouldn't make any choice at all (if it isn't necessary) and instead rely on compiler defaults. But I don't know enough about the reasons behind those defaults, which is why I'm creating an issue instead of a PR.

**Comment by wbhart** [comment]
Flint only compiles as as ANSI C90 library, and we enforce that on all
contributors.

NTL is a C++ library, and the two things are completely independent.
There's no reason to change the default CFLAGS for Flint because NTL
changed something in their library.

**Comment by wbhart** [comment]
Of course, it is possible that Flint compiles the NTL interface in Flint
with the -std=c99 flag, or something like that. This could cause an
incompatibility in the interface itself, which is compiled with a C++
compiler.

That would be a valid issue, and could be resolved by detecting the NTL
version and setting the flag correctly in that instance.

**Comment by timokau** [comment]
> Flint only compiles as as ANSI C90 library, and we enforce that on all contributors.

At least for me it builds and passes the tests with `c++11`.

> NTL is a C++ library, and the two things are completely independent.

But the flint build does fail with the latest ntl and it doesn't when you change the c++ standard.

**Comment by wbhart** [comment]
I can't see how our Makefile could be causing this. The NTL-interface is built using CXXFLAGS, which Flint only sets on BSD, and even then only to C++11.

The CFLAGS should be irrelevant for the NTL-interface as they are only used for the C compiler, not the C++ compiler.

I might be missing something more subtle here, I am not sure. If you think I have overlooked something, please reopen the ticket with an explanation.

**Comment by timokau** [comment]
I don't remember the details of this. I only know that when packaging flint on nixos, we have to set `-std=c++1` since otherwise it will not compile due to ntl. I just re-tried it. This is the error message:

```
    CC   ../build/fq_zech_poly_factor/factor_cantor_zassenhaus.lo
    CC   ../build/fq_zech_poly_factor/is_squarefree.lo
    CC   ../build/fq_zech_poly_factor/../fq_zech_poly_factor.lo
make[2]: Leaving directory '/build/flint-2.5.2/fq_zech_poly_factor'
make[2]: Entering directory '/build/flint-2.5.2'
    CXX   build/interfaces/NTL-interface.lo
In file included from /nix/store/52x4908vr922dhnxz1i5rfnrsq244vzc-gcc-9.3.0/include/c++/9.3.0/type_traits:35,
                 from /nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:25,
                 from /nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/ZZ.h:19,
                 from interfaces/NTL-interface.cpp:32:
/nix/store/52x4908vr922dhnxz1i5rfnrsq244vzc-gcc-9.3.0/include/c++/9.3.0/bits/c++0x_warning.h:32:2: error: #error This file requires compiler and library support for the ISO C++ 2011 standard. This support must be enabled with the -std=c++11 or -std=gnu++11 compiler options.
   32 | #error This file requires compiler and library support \
      |  ^~~~~
In file included from /nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/ZZ.h:19,
                 from interfaces/NTL-interface.cpp:32:
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:20: error: expected ',' or '...' before '&&' token
  664 |    DerivedNode(Args&&... args) : t(std::forward<Args>(args)...) { }
      |                    ^~
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:30: error: parameter packs not expanded with '...':
  664 |    DerivedNode(Args&&... args) : t(std::forward<Args>(args)...) { }
      |                              ^
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:30: note:         'Args'
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h: In constructor 'NTL::details_pthread::DerivedNode<T>::DerivedNode(<type error>)':
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:41: error: 'forward' is not a member of 'std'
  664 |    DerivedNode(Args&&... args) : t(std::forward<Args>(args)...) { }
      |                                         ^~~~~~~
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:41: note: 'std::forward' is only available from C++11 onwards
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:53: error: expected primary-expression before '>' token
  664 |    DerivedNode(Args&&... args) : t(std::forward<Args>(args)...) { }
      |                                                     ^
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:664:55: error: 'args' was not declared in this scope; did you mean 'Args'?
  664 |    DerivedNode(Args&&... args) : t(std::forward<Args>(args)...) { }
      |                                                       ^~~~
      |                                                       Args
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h: At global scope:
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:668:22: error: expected initializer before 'noexcept'
  668 | delete_node(Node *p) noexcept { delete p;  }
      |                      ^~~~~~~~
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h: In function 'void NTL::details_pthread::delete_list(void*)':
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:679:7: error: 'delete_node' was not declared in this scope
  679 |       delete_node(tmp);
      |       ^~~~~~~~~~~
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h: In function 'void NTL::details_pthread::push_node(NTL::details_pthread::Node*)':
/nix/store/bb08kwcw0y6hkq25fx71s18axmzlp7qs-ntl-11.4.3/include/NTL/tools.h:715:7: error: 'delete_node' was not declared in this scope
  715 |       delete_node(p);
      |       ^~~~~~~~~~~
```

it then continues with various similar errors. This is not a big deal for us, since we can just continue setting the `-std=c++11` flag. So if you think that's not a problem on the flint side, I'm fine with keeping this closed.

**Comment by wbhart** [comment]
I'm not really sure. What is CXXFLAGS on your system? Could it be that C++11 is not the default on your system?

**Comment by wbhart** [comment]
Oh, you are using flint-2.5.2, not the trunk branch! This has probably been fixed. Just wait a couple of days and try again with flint-2.6-alpha which is just about to come out.

**Comment by timokau** [comment]
> I'm not really sure. What is CXXFLAGS on your system? Could it be that C++11 is not the default on your system?

Its unset by default and set to `-std=c++11` to fix this issue.

**Comment by timokau** [comment]
> Oh, you are using flint-2.5.2, not the trunk branch! This has probably been fixed. Just wait a couple of days and try again with flint-2.6-alpha which is just about to come out.

Unfortunately not, I verified the issue is still present with latest master (d9d7c5be8dd6f698e700cdb453ff1921bd3f287f).

**Comment by wbhart** [comment]
Ok, then it seems like std-c++11 is not the default for your compiler. There isn't a single version of C++ that supports all versions of NTL and all C++ compilers out there. So I think we will leave this for now, as it would be considerable work to try and accommodate all versions without breaking CXXFLAGS for users.

**Comment by timokau** [comment]
As I mentioned in the original issue, I think this is due to `flint2`'s configure defaulting to `-ansi`, not due to the defaults of my compiler. I have verified this by removing the `-ansi` flag in `flint2`'s source and rebuilding without explicitly passing `-std=c++11`. Removing `-ansi` fixed my issue.

**Comment by wbhart** [comment]
I see the issue now. If CXXFLAGS is not defined by the user, it defaults to using CFLAGS.

I think I can see how to remove ansi from CFLAGS in that case.

**Comment by wbhart** [comment]
I believe I have fixed this in trunk. Let us know if it solves your problem. And thanks for reporting. I wasn't aware that CXXFLAGS was defaulting to CFLAGS, so it is good to have this -ansi flag thing fixed.

**Comment by timokau** [comment]
Yes, I can confirm the issue is fixed on c6319d1d36248f2fc699e833ab2f6fa70d21e906 :tada:

Thank you!

**Comment by wbhart** [comment]
Thanks for the testing and feedback!
